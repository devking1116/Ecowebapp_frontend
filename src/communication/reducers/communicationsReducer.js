/* @flow */
import { COMMUNICATION_REDUCER_LOAD } from '../actions/loadCommunicationsAction';
import { COMMUNICATION_REDUCER_SELECT } from '../actions/selectCommunicationAction';

export type Communication = {
  title: string,
  author: string,
  date: string,
  status?: 'active' | 'in-active',
  attachments?: Array<{
    id: number,
    fileName: string,
    size: string,
  }>,
  comments?: Array<{
    author: string,
    date: string,
    status?: 'active' | 'in-active',
    actionsTakens: Array<string>,
    linkedLogbooks: Array<string>,
    attachments?: Array<number>,
  }>,
  activity?: Array<{
    author: string,
    label: string,
    date: string,
  }>,
};

type State = {
  communications: Array<Communication>,
  selectedCommunication: ?Communication,
};

const initialState = {
  communications: [],
  selectedCommunication: null,
};

/**
 * The reducer that contains all the communication data for the site.
 */
export const COMMUNICATION_REDUCER_STATE = 'COMMUNICATION_REDUCER_STATE';
export default function communicationsReducer(state: State = initialState, action: Object) {
  switch (action.type) {
    case COMMUNICATION_REDUCER_LOAD:
      //TODO, replace hard-coded data with data from the api
      return {
        ...state,
        communications: [
          { title: 'Testing departments acess', author: 'User 0', date: '14/Dec/2016 00:21 PM MST', status: 'active' },
          { title: 'Test communication 1', author: 'User 1', date: '14/Dec/2016 01:21 PM MST' },
          { title: 'Test communication 2', author: 'User 2', date: '14/Dec/2016 01:21 PM MST', status: 'in-active' },
        ],
      };
    case COMMUNICATION_REDUCER_SELECT:
      //TODO, replace hard-coded data with data from the api
      return {
        ...state,
        selectedCommunication: {
          ...state.communications[action.index],
          attachments: [
            { id: 0, fileName: 'testAttachment0.pdf', modifiedDate: 'January 3, 2017', size: '67.0kb' },
            { id: 1, fileName: 'testAttachment1.pdf', modifiedDate: 'January 4, 2016', size: '67.1kb' },
            { id: 2, fileName: 'testAttachment2.pdf', modifiedDate: 'January 7, 2017', size: '67.2kb' },
          ],
          comments: [
            {
              author: 'Richard Simms',
              date: 'Today 5:45 am',
              body: 'Test comment 0',
              status: 'active',
              actionsTakens: ['Spoke to Resident', 'Jetted Sweer Line'],
              linkedLogbooks: [ 'logbook A' ],
            },
            {
              author: 'Randy S.Beard',
              date: '1/1/2016',
              body: 'Test comment 1',
              linkedLogbooks: [ 'logbook A', 'logbook B' ],
              attachments: [ 0 ],
            },
            {
              author: 'User 2',
              date: '2/2/2016',
              body: 'Test comment 2',
            },
            {
              author: 'User 3',
              date: '3/3/2016',
              body: 'Test comment 3',
              status: 'in-active',
              attachments: [ 1, 2 ],
            },
          ],
          activity: [
            { author: 'User 0', date: '14/Dec/2016 00:21 PM MST', label: 'commented on this post' },
            { author: 'User 1', date: '14/Dec/2016 01:21 PM MST', label: 'added two attachments' },
            { author: 'User 2', date: '14/Dec/2016 02:21 PM MST', label: 'change status to active' },
          ],
        },
      };
    default:
      return state;
  }
}
