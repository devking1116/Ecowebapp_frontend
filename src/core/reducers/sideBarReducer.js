/* @flow */
import { SIDEBAR_COLLAPSE } from '../actions/collapseSideBarAction';
import { SIDEBAR_EXPAND } from '../actions/expandSideBarAction';
import { SIDEBAR_REGISTER } from '../actions/registerSideBarAction';
import { SIDEBAR_UNREGISTER } from '../actions/unregisterSideBarAction';

export type SideBarState = {
  expandedWidth: number | string,
  isExpanded: bool,
};

export type State = { [key: string]: SideBarState };

/**
 * A reducer to control weather or not a component is expaned or collapsed.
 */
export const SIDEBAR_REDUCER_STATE = 'SIDEBAR_REDUCER_STATE';
export default function sideBarReducer(state: State = { }, action: Object): State {
  switch(action.type) {
    case SIDEBAR_REGISTER:
      if(!action.sideBarId) {
        throw new Error('Side Bar must be given to unregister');
      }
      if(typeof action.expandedWidth !== 'number' && typeof action.expandedWidth !== 'string') {
        throw new Error('Side Bar expandedWidth must be number or string');
      }

      return {
        ...state,
        [action.sideBarId]: {
          expandedWidth: action.expandedWidth,
          isExpanded: action.defaultExpanded? true: false,
        },
      };
    case SIDEBAR_UNREGISTER:
      if(!action.sideBarId)         throw new Error('Side Bar must be given to unregister');

      return {
        ...state,
        [action.sideBarId]: undefined,
      };
    case SIDEBAR_EXPAND:
    case SIDEBAR_COLLAPSE:
      if(!action.sideBarId)         throw new Error('Side Bar must be given to update');
      if(!state[action.sideBarId])  throw new Error('Side Bar must be registered to update');

      return {
        ...state,
        [action.sideBarId]: {
          ...state[action.sideBarId],
          isExpanded: action.type === SIDEBAR_EXPAND? true: false,
        },
      };
    default:
      return state;
  }
}
