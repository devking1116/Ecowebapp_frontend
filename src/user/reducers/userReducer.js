/* @flow */
import { USER_REDUCER_LOGIN } from '../actions/loginUserAction';

type State = {
  clientName: ?string,
  clientId: ?string,
  isAdmin: bool,
};

const initialState = {
  clientName: null,
  clientId: null,
  isAdmin: false,
};

/**
 * A reducer to keep track of the data for the logged in user.
 */
export const USER_REDUCER_STATE = 'USER_REDUCER_STATE';
export default function userReducer(state: State = initialState, action: *): State {
  switch (action.type) {
    case USER_REDUCER_LOGIN:
      //TODO, replace hard-coded data with data from the api
      return {
        clientName: 'Client Name',
        clientId:'123456',
        isAdmin: true,
      };
    default:
      return state;
  }
}
