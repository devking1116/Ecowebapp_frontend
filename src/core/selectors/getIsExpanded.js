/* @flow */
import { SIDEBAR_REDUCER_STATE } from '../reducers/sideBarReducer';

import type { SideBarState } from '../reducers/sideBarReducer';

/**
 * Selects the isExpanded prop from the reducer state,
 */
export default function getIsExpanded(state: Object, sideBarId: string): SideBarState {
  const sidebarStates = state[SIDEBAR_REDUCER_STATE];
  if(!sidebarStates) throw new Error(`Missing required reducer: ${SIDEBAR_REDUCER_STATE}`);

  if(sidebarStates[sideBarId])  return sidebarStates[sideBarId];
  else                          return { isExpanded: false, expandedWidth: '100%' };
}
