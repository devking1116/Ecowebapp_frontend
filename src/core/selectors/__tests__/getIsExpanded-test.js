import { SIDEBAR_REDUCER_STATE } from '../../reducers/sideBarReducer';
import getIsExpanded from '../getIsExpanded';

it('adds default isExpanded / expandedWidth values when the current sidebar has not been registed', () => {
  const TEST_SIDEBAR = 'TEST_SIDEBAR';
  const sideBarState = getIsExpanded({ [SIDEBAR_REDUCER_STATE]: { } }, TEST_SIDEBAR);

  expect(sideBarState.isExpanded).toBe(false);
  expect(sideBarState.expandedWidth).toEqual('100%');
});

it('get the isExpanded / expandedWidth values from the state', () => {
  const TEST_SIDEBAR = 'TEST_SIDEBAR';
  const isExpanded = true;
  const width = 200;
  const state = { [TEST_SIDEBAR]: { isExpanded, expandedWidth: width } };
  const sideBarState = getIsExpanded({ [SIDEBAR_REDUCER_STATE]: state }, TEST_SIDEBAR);

  expect(sideBarState.isExpanded).toEqual(isExpanded);
  expect(sideBarState.expandedWidth).toEqual(width);
});

it('throws an error if the reducer is not in the state', () => {
  const callWithOutReducer = () => getIsExpanded({ }, '');

  expect(callWithOutReducer).toThrowErrorMatchingSnapshot();
});
