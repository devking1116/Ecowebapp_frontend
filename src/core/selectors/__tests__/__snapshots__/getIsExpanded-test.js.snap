exports[`test throws an error if the reducer is not in the state 1`] = `"Missing required reducer: SIDEBAR_REDUCER_STATE"`;
